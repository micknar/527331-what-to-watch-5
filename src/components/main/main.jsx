import React from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import FilmsList from "../films-list/films-list";
import Genres from "../genres/genres";
import ShowMoreBtn from "../show-more-btn/show-more-btn";
import PageHeaderLogo from "../page-header-logo/page-header-logo";
import UserBlock from "../user-block/user-block";
import PageFooter from "../page-footer/page-footer";
import AddToFavoriteBtn from '../add-to-favorite-btn/add-to-favorite-btn';
import withActiveCard from "../../hocs/with-active-card/with-active-card";
import {getFilteredFilms} from "../../store/reducers/app-state/selectors";
import {AppRoute} from "../../const";

const FilmsListWrapped = withActiveCard(FilmsList);

const Main = (props) => {
  const {promoFilm, filteredFilms, renderedFilmsCount, onFilmCardClick} = props;
  const {posterImage, name, genre, backgroundImage, backgroundColor, released, id, isFavorite} = promoFilm;

  return (
    <>
      <section className="movie-card" style={{backgroundColor}}>
        <div className="movie-card__bg">
          <img src={backgroundImage} alt={name} />
        </div>

        <h1 className="visually-hidden">WTW</h1>

        <header className="page-header movie-card__head">
          <PageHeaderLogo />
          <UserBlock />
        </header>

        <div className="movie-card__wrap">
          <div className="movie-card__info">
            <div className="movie-card__poster">
              <img src={posterImage} alt={name} width="218" height="327" />
            </div>

            <div className="movie-card__desc">
              <h2 className="movie-card__title">{name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{genre}</span>
                <span className="movie-card__year">{released}</span>
              </p>

              <div className="movie-card__buttons">
                <Link to={AppRoute.PLAYER + id} className="btn btn--play movie-card__button">
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </Link>

                <AddToFavoriteBtn
                  id={id}
                  isFavorite={isFavorite}
                />
              </div>
            </div>
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>

          <Genres />

          <FilmsListWrapped
            films={filteredFilms.slice(0, renderedFilmsCount)}
            onFilmCardClick={onFilmCardClick}
          />

          {renderedFilmsCount < filteredFilms.length && <ShowMoreBtn />}
        </section>

        <PageFooter />
      </div>
    </>
  );
};

const mapStateToProps = ({APP_STATE}) => ({
  filteredFilms: getFilteredFilms({APP_STATE}),
  renderedFilmsCount: APP_STATE.renderedFilmsCount,
  promoFilm: APP_STATE.promoFilm,
});

Main.propTypes = {
  promoFilm: PropTypes.shape({
    id: PropTypes.number.isRequired,
    posterImage: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    backgroundImage: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
    released: PropTypes.number.isRequired,
    isFavorite: PropTypes.bool.isRequired,
  }).isRequired,

  filteredFilms: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    previewVideoLink: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
  })).isRequired,

  renderedFilmsCount: PropTypes.number.isRequired,
  onFilmCardClick: PropTypes.func.isRequired,
};

export {Main};
export default connect(mapStateToProps)(Main);
