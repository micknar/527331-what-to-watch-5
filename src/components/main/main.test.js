import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {Main} from "./main";
import {noop, films} from "../../test-mocks";

jest.mock(`../films-list/films-list`, () => `FilmsListWrapped`);
jest.mock(`../genres/genres`, () => `Genres`);
jest.mock(`../show-more-btn/show-more-btn`, () => `ShowMoreBtn`);
jest.mock(`../page-header-logo/page-header-logo`, () => `PageHeaderLogo`);
jest.mock(`../user-block/user-block`, () => `UserBlock`);
jest.mock(`../page-footer/page-footer`, () => `PageFooter`);
jest.mock(`../add-to-favorite-btn/add-to-favorite-btn`, () => `AddToFavoriteBtn`);

it(`Should Main render correctly`, () => {
  const tree = renderer
    .create(
        <BrowserRouter>
          <Main
            promoFilm={films[0]}
            filteredFilms={films}
            renderedFilmsCount={8}
            onCardClick={noop}
          />
        </BrowserRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
