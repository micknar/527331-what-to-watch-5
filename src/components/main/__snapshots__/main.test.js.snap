// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Main render correctly 1`] = `
Array [
  <section
    className="movie-card"
    style={
      Object {
        "backgroundColor": "#4aa2e5",
      }
    }
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Bronson"
        src="/img/the-grand-budapest-hotel-bg.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <PageHeaderLogo />
      <UserBlock />
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="Bronson"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Bronson
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Crime
            </span>
            <span
              className="movie-card__year"
            >
              1996
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/1"
              onClick={[Function]}
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <AddToFavoriteBtn
              id={1}
              isFavorite={false}
            />
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <Genres />
      <FilmsListWrapped
        activeCard={-1}
        films={
          Array [
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "/img/the-grand-budapest-hotel-bg.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
              "director": "Nicolas Winding Refn",
              "genre": "Crime",
              "id": 1,
              "isFavorite": false,
              "name": "Bronson",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 8.9,
              "ratingMark": "Very good",
              "released": 1996,
              "runTime": Object {
                "hours": 2,
                "minutes": 40,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "director": "Quentin Tarantino",
              "genre": "Thriller",
              "id": 2,
              "isFavorite": true,
              "name": "Midnight Special",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Midnight_Special.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/Midnight_Special.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 10,
              "ratingMark": "Awesome",
              "released": 1989,
              "runTime": Object {
                "hours": 1,
                "minutes": 42,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem ipsum dolor sit amo eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "director": "Quentin Tarantino",
              "genre": "Action",
              "id": 3,
              "isFavorite": false,
              "name": "Pulp Fiction",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/Pulp_Fiction.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 2.2,
              "ratingMark": "Bad",
              "released": 2007,
              "runTime": Object {
                "hours": 1,
                "minutes": 22,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ext enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "director": "Quentin Tarantino",
              "genre": "Adventure",
              "id": 4,
              "isFavorite": true,
              "name": "Matrix",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Matrix.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/Matrix.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 4.5,
              "ratingMark": "Normal",
              "released": 2011,
              "runTime": Object {
                "hours": 1,
                "minutes": 24,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem i ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "director": "Quentin Tarantino",
              "genre": "Comedy",
              "id": 5,
              "isFavorite": false,
              "name": "Legend",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Legend.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/Legend.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 4.1,
              "ratingMark": "Normal",
              "released": 2015,
              "runTime": Object {
                "hours": 1,
                "minutes": 57,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sedrem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.uip ex ea commodo consequat.",
              "director": "Quentin Tarantino",
              "genre": "Fantasy",
              "id": 6,
              "isFavorite": false,
              "name": "angs of new york",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/gangs_of_new_york.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/gangs_of_new_york.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 7.8,
              "ratingMark": "Good",
              "released": 1980,
              "runTime": Object {
                "hours": 1,
                "minutes": 35,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "director": "Guy Ritchie",
              "genre": "Drama",
              "id": 7,
              "isFavorite": false,
              "name": "A Star Is Born",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/A_Star_is_Born.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/A_Star_is_Born.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 6.4,
              "ratingMark": "Good",
              "released": 1989,
              "runTime": Object {
                "hours": 1,
                "minutes": 25,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Bohemian_Rhapsody.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliqliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
              "director": "Quentin Tarantino",
              "genre": "Fantasy",
              "id": 8,
              "isFavorite": false,
              "name": "Snatch",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/Snatch.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 9.8,
              "ratingMark": "Very good",
              "released": 2004,
              "runTime": Object {
                "hours": 2,
                "minutes": 20,
              },
              "scoresCount": 240,
              "starring": "Bill Murray, Edward Norton, Jude Law, Willem Dafoe, Saoirse Ronan",
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            },
          ]
        }
        onCardClick={[Function]}
        onCardMouseOut={[Function]}
        onCardMouseOver={[Function]}
      />
      <ShowMoreBtn />
    </section>
    <PageFooter />
  </div>,
]
`;
